fix work queue polling algorithm

okummer opened the pull request:
If the old implementation received a notification for a new work
item when it was about to awake from its wait() call anyway, it
incorrectly assumed that the work queue was empty. Subsequently,
it did not fetch an item from the work queue and terminated the
current worker thread.

This might terminate the last worker thread, keeping work items
in the queue forever. Because the work items might be tasked with
reenabling the selector thread, it might have permantently
rendered an ORB unresponsive.

The new implementation checks the work queue directly instead
of relying on timing information. It also continues waiting if
the wait() call returned early, thereby guaranteeing that the
wait timeout is exhausted before terminating the thread. In the
rare case of a System clock that jumps backward in time, it makes
sure not to keep the threads alive for too long.

The upfront check for an empty queue outside of the try statement
has been removed to keep the code simpler.

This fixes https://github.com/eclipse-ee4j/orb/issues/18 .

Signed-off-by: Olaf Kummer <olaf.kummer@coremedia.com>

russgold commented:
Are you able to create a unit test that I can run against the old version and verify that it fails, and that passes with the new one? If not, can you explain your fix so that I can create such a test? It looks as though you've made a lot of changes.

okummer commented:
I have pushed a second commit containing the reproducer for bug 18 and also a set of basic unit tests for the thread pool.

Of course, you are right that I made quite a few changes. Let me summarize the reasons:
* I added a waiting loop instead of a single call to wait() to get more in line with best concurrent programming practices.
* Having a loop meant that I had to a separate conditional for checking whether the queue is empty after waiting.
* That made it possible to collapse it with the if that checks whether the work item is null, making it more clear that we have two distinct paths.
* The outer if surrounding the try had by then become redundant, so I removed it for code clarity. This changed the indentation, making it more difficult for git to spot similarities.
* I flipped the condition work!=null to work==null, because positive conditions are generally easier to understand.
* Finally, I inlined the method isQueueEmpty(), principally because it was used only once and because its name was not helping.

Thanks for looking into this so quickly! Are there any other questions about this fix that I can help with?

okummer commented:
Can I assist in the handling of this PR in any way?

russgold commented:
Sorry for the delay in getting back to this; I am in the critical path of way too many projects. My concern currently is that the unit tests (thanks for them, BTW) are very slow. There are techniques for simulating delays such as you are using, and my intention was to apply them before merging. If you know how, you could do it yourself; otherwise, I will prioritize doing it in the next few days.

okummer commented:
Thanks for your time and effort! You are right about the test taking a long time. I just saw no way of bringing that down without adding hook methods in the thread pool and the work queue and interfering with the setup of the work queue, which is currently hardwired in the thread pool. All of that would have blown up the changes to the core classes even more...
